text/0000-inherent-trait-impl.md

```diff
@@ -0,0 +1,152 @@
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Provides a mechanism to declare "inherent traits" for a type defined in the same crate. Methods on these traits are callable on instances of the specified type without needing to import the trait.
+
+# Motivation
+[motivation]: #motivation
+
+There are two similar cases where this is valuable:
+
+- Frequently used traits.
+
+  Sometimes getting the right abstractions require breaking up a type's implementation into many traits, with only a few methods per
+  trait. Every use of such a type results in a large number of imports to ensure the correct traits are in scope. If such a type is used
+  frequently, then this burden quickly becomes a pain point for users of the API,
+  especially if users do not care about writing generic code over traits.
+
+- Mapping object-oriented APIs.
+
+  When mapping these APIs to rust, base classes are usually mapped to traits: methods on those base classes will need to be callable on any
+  derived type. This is sub-optimal because to use a class method a user must now know which class in the hierachy defined that
+  method, so that they can import and use the corresponding trait. This knowledge is not required when using the same API from an
+  object-oriented language.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The feature is implemented using a new attribute which can be applied to trait
+`impl` blocks:
+
+```rust
+pub trait Bar {
+    fn bar(&self);
+}
+
+pub trait ExtBar: Bar {
+    fn ext_bar(&self);
+}
+
+impl<T: Bar> ExtBar for T {
+    fn ext_bar(&self) { }
+}
+
+pub struct Foo;
+
+#[inherent]
+impl Bar for Foo {
+    fn bar(&self) { println!("foo::bar"); }
+}
+
+// works thanks to specialization
+#[inherent]
+impl ExtBar for Foo { }
+
+impl Foo {
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+The methods `bar` and `ext_bar` are now callable on any instance of `Foo`,
+regardless of whether the `Bar` and `ExtBar` traits are currently in scope,
+or even whether the `Bar` trait is publically visible. In other words if `Bar`
+and `ExtBar` defined in one crate and `Foo` and another, user of `Foo` will be
+able to explicitly depend only on the crate which defines `Foo` and still use
+the inherent traits methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `inherent` attribute in the above example makes the `impl` block equivalent to:
+
+```rust
+impl Foo {
+    #[inline]
+    pub fn bar(&self) { <Self as Bar>::bar(self); }
+
+    fn foo(&self) { println!("foo::foo"); }
+}
+```
+
+Any questions regarding coherence, visibility or syntax can be resolved by
+comparing against this expansion, although the feature need not be implemented
+as an expansion within the compiler.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Increased complexity of the language.
+- Hides use of traits from users.
+
```

Comment:
The older RFC was closed because it may be preferable to do a general purpose delegation RFC. This should be mentioned and preferably addressed.

https://github.com/rust-lang/rfcs/pull/2309#issuecomment-369731403

Reply:
Firstly, I think that delegation and inherent impls are orthogonal to each other. Secondly, IIRC the functionality which may remove the need for inherent impls is listed as potential future extension of delegation RFC, i.e. the current delegation RFC does not solve the target issue of this proposal. And thirdly, I believe both proposals should be reviewed in parallel and pros and cons of both proposals should be weighted and compared, so I don't think "it may be preferable" is a good argument for closing competing proposals.

Reply:
I do like inherent trait impls, but they are only a very minor paper cut, while delegation is extremely important, after say ATCs and const generics.  Inherent trait impls are simply not worth complicating delegation though, so maybe the previous postponement should stand until the delegation conversation happens.

As an aside, I'd worry far more about the state of doc comments on trait impls than about inherent trait impls, as poorly documented impls create more headaches.  I donno if https://github.com/rust-lang/rust/issues/18701 ever got fixed, but if so then this may only be cultural now.

Reply:
At this point, I think it'd be reasonable to pursue inherent traits, and I don't think doing so would block delegation in any meaningful way. Having `#[inherent] impl Trait for Type` doesn't make it any harder to implement delegation.